File: src/ados/plugins/msp.py
"""MSP (MultiWii Serial Protocol) v1 + v2 codec — the Python twin of
``crates/ados-protocol/src/msp.rs``.

The agent's router exposes a raw MSP byte pipe on ``/run/ados/msp.sock`` to a
Betaflight / iNav / KISS flight controller, but the Rust codec that builds and
parses MSP frames is only reachable from a Rust plugin. A Python companion loop
that flies an MSP FC (``MSP_SET_RAW_RC``) needs the same codec on this side, so
this module ports it byte-for-byte: the same two wire formats, the same CRC, and
the same stick scaling the GCS standardized. The parity tests pin the identical
golden vectors the Rust module pins, so the two halves cannot drift on the wire.

Two frame formats, both little-endian:

* **MSPv1** ``$M<``: ``'$' 'M' <dir> <size:u8> <cmd:u8> <payload...> <crc:u8>``,
  ``crc`` = XOR of ``size``, ``cmd``, and every payload byte.
* **MSPv2** ``$X<``: ``'$' 'X' <dir> <flag:u8> <cmd:u16 LE> <size:u16 LE>
  <payload...> <crc:u8>``, ``crc`` = CRC8/DVB-S2 (poly ``0xD5``) over ``flag``,
  the two command bytes, the two size bytes, and every payload byte.
"""

from __future__ import annotations

import math
from dataclasses import dataclass

# MSP_SET_RAW_RC — write raw RC channel values (PWM microseconds) to the FC.
# Payload is N little-endian u16 channels. This is what flies an FC in a rate
# mode (ACRO), the FC treating the values as if they came from a receiver.
MSP_SET_RAW_RC = 200

DIR_TO_FC = ord("<")
DIR_FROM_FC = ord(">")
DIR_ERROR = ord("!")


class MspDecodeError(Exception):
    """Base for the reasons a buffer is not a single complete MSP frame."""


class MspTooShort(MspDecodeError):
    """Fewer bytes than a complete frame of its kind — wait for more."""


class MspBadPreamble(MspDecodeError):
    """The ``$M`` / ``$X`` preamble or version byte was not recognized."""


class MspBadCrc(MspDecodeError):
    """A complete frame whose trailing CRC did not match (corrupt)."""

    def __init__(self, expected: int, found: int) -> None:
        super().__init__(f"bad crc: expected {expected:#04x}, found {found:#04x}")
        self.expected = expected
        self.found = found


def crc8_dvb_s2(crc: int, byte: int) -> int:
    """Fold one byte into a CRC8/DVB-S2 running value (polynomial ``0xD5``)."""
    crc ^= byte
    for _ in range(8):
        if crc & 0x80:
            crc = ((crc << 1) ^ 0xD5) & 0xFF
        else:
            crc = (crc << 1) & 0xFF
    return crc


def crc8_dvb_s2_buf(data: bytes) -> int:
    """CRC8/DVB-S2 over a buffer, seeded at 0."""
    crc = 0
    for b in data:
        crc = crc8_dvb_s2(crc, b)
    return crc


def encode_v1(cmd: int, payload: bytes) -> bytes | None:
    """Encode an MSPv1 request ``$M<``. Returns ``None`` when ``cmd`` or the
    payload length exceeds a byte (a v2-only frame) rather than truncating."""
    if cmd > 0xFF or len(payload) > 0xFF:
        return None
    size = len(payload)
    out = bytearray([ord("$"), ord("M"), DIR_TO_FC, size, cmd])
    out.extend(payload)
    crc = size ^ cmd
    for b in payload:
        crc ^= b
    out.append(crc & 0xFF)
    return bytes(out)


def encode_v2(cmd: int, payload: bytes) -> bytes:
    """Encode an MSPv2 request ``$X<`` with ``flag = 0`` (the common case)."""
    return encode_v2_flagged(0, cmd, payload)


def encode_v2_flagged(flag: int, cmd: int, payload: bytes) -> bytes:
    """Encode an MSPv2 request ``$X<`` with an explicit flag byte."""
    size = len(payload)
    header = bytes([flag & 0xFF]) + cmd.to_bytes(2, "little") + size.to_bytes(2, "little")
    out = bytearray([ord("$"), ord("X"), DIR_TO_FC])
    out.extend(header)
    out.extend(payload)
    crc = crc8_dvb_s2_buf(header)
    for b in payload:
        crc = crc8_dvb_s2(crc, b)
    out.append(crc)
    return bytes(out)


@dataclass(frozen=True)
class MspFrame:
    """A decoded MSP frame: command id, direction/marker byte, payload, version."""

    cmd: int
    dir: int
    payload: bytes
    version: int


def decode_frame(buf: bytes) -> tuple[MspFrame, int]:
    """Decode exactly one MSP frame from the START of ``buf`` (v1 or v2,
    auto-detected). Returns ``(frame, consumed)`` so a streaming caller can
    advance. Raises :class:`MspTooShort` / :class:`MspBadPreamble` /
    :class:`MspBadCrc`; fails closed on a CRC mismatch (Rule 44)."""
    if len(buf) < 3:
        raise MspTooShort
    if buf[0] != ord("$"):
        raise MspBadPreamble
    if buf[1] == ord("M"):
        return _decode_v1(buf)
    if buf[1] == ord("X"):
        return _decode_v2(buf)
    raise MspBadPreamble


def _decode_v1(buf: bytes) -> tuple[MspFrame, int]:
    if len(buf) < 6:
        raise MspTooShort
    direction = buf[2]
    size = buf[3]
    cmd = buf[4]
    total = 6 + size
    if len(buf) < total:
        raise MspTooShort
    payload = bytes(buf[5 : 5 + size])
    crc = buf[3] ^ buf[4]
    for b in payload:
        crc ^= b
    found = buf[total - 1]
    if crc != found:
        raise MspBadCrc(crc, found)
    return MspFrame(cmd=cmd, dir=direction, payload=payload, version=1), total


def _decode_v2(buf: bytes) -> tuple[MspFrame, int]:
    if len(buf) < 9:
        raise MspTooShort
    direction = buf[2]
    cmd = int.from_bytes(buf[4:6], "little")
    size = int.from_bytes(buf[6:8], "little")
    total = 9 + size
    if len(buf) < total:
        raise MspTooShort
    payload = bytes(buf[8 : 8 + size])
    crc = crc8_dvb_s2_buf(bytes(buf[3:8]))  # flag + cmd + size
    for b in payload:
        crc = crc8_dvb_s2(crc, b)
    found = buf[total - 1]
    if crc != found:
        raise MspBadCrc(crc, found)
    return MspFrame(cmd=cmd, dir=direction, payload=payload, version=2), total


def set_raw_rc(channels: list[int], v2: bool = True) -> bytes | None:
    """Encode ``MSP_SET_RAW_RC`` carrying ``channels`` PWM values as little-endian
    ``u16``s. ``None`` only for a v1 encode with too many channels."""
    payload = b"".join(int(ch).to_bytes(2, "little") for ch in channels)
    if v2:
        return encode_v2(MSP_SET_RAW_RC, payload)
    return encode_v1(MSP_SET_RAW_RC, payload)


def _round_half_away(x: float) -> int:
    """Round half away from zero, matching Rust ``f32::round`` (Python's built-in
    ``round`` is banker's rounding, which would differ on a ``.5`` boundary)."""
    return int(math.floor(x + 0.5)) if x >= 0 else int(math.ceil(x - 0.5))


def bipolar_to_pwm(v: float) -> int:
    """Map a bipolar stick axis (roll/pitch/yaw), ``-1.0..=1.0``, to PWM
    ``1000..=2000`` with centre ``1500``. Out-of-range clamps."""
    clamped = max(-1.0, min(1.0, v))
    return _round_half_away(1500.0 + clamped * 500.0)


def throttle_to_pwm(t: float) -> int:
    """Map throttle ``0.0..=1.0`` to PWM ``1000..=2000`` with idle at ``1000``
    (NOT centre — a throttle at rest is idle, not mid-stick). Clamps."""
    clamped = max(0.0, min(1.0, t))
    return _round_half_away(1000.0 + clamped * 1000.0)


def sticks_to_channels(roll: float, pitch: float, yaw: float, throttle: float) -> list[int]:
    """Pack normalized sticks into AETR PWM channels ``[roll, pitch, throttle,
    yaw]`` — the codebase convention. Mirrors the Rust SDK ``send_sticks``."""
    return [
        bipolar_to_pwm(roll),
        bipolar_to_pwm(pitch),
        throttle_to_pwm(throttle),
        bipolar_to_pwm(yaw),
    ]


# A real MSP frame cannot exceed the v2 header + a u16 payload + CRC.
_MAX_MSP_FRAME = 9 + 0xFFFF + 1


class MspReassembler:
    """Accumulate a delivered MSP byte stream and emit whole frames — the Python
    twin of the SDK reassembler. A companion reading FC responses off the pipe
    feeds each delivered chunk to :meth:`push`; a frame straddling two chunks is
    reassembled rather than dropped, a corrupt frame is skipped, and a
    preamble-less stream is drained rather than hoarded."""

    def __init__(self) -> None:
        self._buf = bytearray()

    def push(self, chunk: bytes) -> list[MspFrame]:
        """Append ``chunk`` and return every complete frame now decodable."""
        self._buf.extend(chunk)
        out: list[MspFrame] = []
        while True:
            try:
                frame, consumed = decode_frame(bytes(self._buf))
            except MspTooShort:
                if len(self._buf) > _MAX_MSP_FRAME:
                    self._resync()
                    continue
                break
            except (MspBadPreamble, MspBadCrc):
                self._resync()
                continue
            out.append(frame)
            del self._buf[:consumed]
        return out

    @property
    def pending(self) -> int:
        """Bytes buffered but not yet a complete frame."""
        return len(self._buf)

    def _resync(self) -> None:
        """Advance to the next ``$``; clear the buffer if there is none. Always
        shrinks, so :meth:`push`'s loop cannot spin."""
        nxt = self._buf.find(b"$", 1)
        if nxt == -1:
            self._buf.clear()
        else:
            del self._buf[:nxt]

File: crates/ados-protocol/src/msp.rs
//! MSP (MultiWii Serial Protocol) v1 + v2 codec.
//!
//! The agent's MAVLink router already exposes a raw MSP byte pipe on
//! `/run/ados/msp.sock` (it forwards bytes to and from a Betaflight / iNav /
//! KISS flight controller's serial link), but nothing in the tree could build
//! or parse an MSP frame. This module is that codec: the wire framing for both
//! MSP versions plus the one command a companion needs to fly a rate-control FC,
//! `MSP_SET_RAW_RC` (200), with the stick scaling the GCS already standardized.
//!
//! Two frame formats, both little-endian:
//!
//! * **MSPv1** `$M<`: `'$' 'M' <dir> <size:u8> <cmd:u8> <payload...> <crc:u8>`
//!   where `crc` is the XOR of `size`, `cmd`, and every payload byte. `size` is
//!   the payload length, so a v1 command id and a single payload byte cannot
//!   exceed 255. `<dir>` is `<` for a request to the FC, `>` for a reply, `!`
//!   for an error reply.
//! * **MSPv2** `$X<`: `'$' 'X' <dir> <flag:u8> <cmd:u16 LE> <size:u16 LE>
//!   <payload...> <crc:u8>` where `crc` is CRC8/DVB-S2 (poly `0xD5`) accumulated
//!   over `flag`, the two command bytes, the two size bytes, and every payload
//!   byte. v2 lifts the 8-bit command/size ceiling (needed for the full command
//!   space and larger settings payloads).
//!
//! The codec is transport-agnostic: it produces and consumes byte buffers, and
//! the plugin host / MSP client move those over the socket. It never opens the
//! socket or the FC itself.

/// `MSP_SET_RAW_RC` — write raw RC channel values (PWM microseconds) to the FC.
/// Payload is N little-endian `u16` channels. This is the command a companion
/// uses to fly an FC in a rate mode (ACRO), the FC treating the values as if
/// they came from a receiver.
pub const MSP_SET_RAW_RC: u16 = 200;

/// The v1 direction/marker byte for a request to the FC (`<`).
pub const DIR_TO_FC: u8 = b'<';
/// The v1 marker for a normal reply from the FC (`>`).
pub const DIR_FROM_FC: u8 = b'>';
/// The v1 marker for an error reply from the FC (`!`).
pub const DIR_ERROR: u8 = b'!';

/// Accumulate one byte into a CRC8/DVB-S2 running value (polynomial `0xD5`).
/// Seed with `0` and fold every byte of the v2 header+payload.
pub fn crc8_dvb_s2(mut crc: u8, byte: u8) -> u8 {
    crc ^= byte;
    for _ in 0..8 {
        if crc & 0x80 != 0 {
            crc = (crc << 1) ^ 0xD5;
        } else {
            crc <<= 1;
        }
    }
    crc
}

/// CRC8/DVB-S2 over a slice, seeded at 0.
pub fn crc8_dvb_s2_buf(data: &[u8]) -> u8 {
    data.iter().fold(0u8, |c, &b| crc8_dvb_s2(c, b))
}

/// Encode an MSPv1 request `$M<`. `cmd` and `payload.len()` must each fit in a
/// byte; a command or payload that does not is a v2-only frame, so this returns
/// `None` rather than truncating (a silent truncation would corrupt the FC
/// command — fail loudly instead).
pub fn encode_v1(cmd: u16, payload: &[u8]) -> Option<Vec<u8>> {
    if cmd > u8::MAX as u16 || payload.len() > u8::MAX as usize {
        return None;
    }
    let size = payload.len() as u8;
    let cmd = cmd as u8;
    let mut out = Vec::with_capacity(6 + payload.len());
    out.extend_from_slice(&[b'$', b'M', DIR_TO_FC, size, cmd]);
    out.extend_from_slice(payload);
    let mut crc = size ^ cmd;
    for &b in payload {
        crc ^= b;
    }
    out.push(crc);
    Some(out)
}

/// Encode an MSPv2 request `$X<` with `flag = 0` (the common case).
pub fn encode_v2(cmd: u16, payload: &[u8]) -> Vec<u8> {
    encode_v2_flagged(0, cmd, payload)
}

/// Encode an MSPv2 request `$X<` with an explicit flag byte.
pub fn encode_v2_flagged(flag: u8, cmd: u16, payload: &[u8]) -> Vec<u8> {
    let size = payload.len() as u16;
    let mut out = Vec::with_capacity(9 + payload.len());
    out.extend_from_slice(&[b'$', b'X', DIR_TO_FC]);
    // The CRC covers flag, cmd (LE), size (LE), and the payload — everything
    // after the direction byte except the CRC itself.
    let mut header = Vec::with_capacity(5);
    header.push(flag);
    header.extend_from_slice(&cmd.to_le_bytes());
    header.extend_from_slice(&size.to_le_bytes());
    out.extend_from_slice(&header);
    out.extend_from_slice(payload);
    let mut crc = crc8_dvb_s2_buf(&header);
    for &b in payload {
        crc = crc8_dvb_s2(crc, b);
    }
    out.push(crc);
    out
}

/// A decoded MSP frame: its command id, direction/marker byte, and payload.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MspFrame {
    pub cmd: u16,
    /// The direction byte: [`DIR_TO_FC`], [`DIR_FROM_FC`], or [`DIR_ERROR`].
    pub dir: u8,
    pub payload: Vec<u8>,
    /// The MSP version the frame was encoded in (1 or 2).
    pub version: u8,
}

/// Why a byte buffer could not be decoded as a single complete MSP frame.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum MspDecodeError {
    /// Fewer bytes than the smallest valid frame of its kind.
    TooShort,
    /// The `$M` / `$X` preamble or the version byte was not recognized.
    BadPreamble,
    /// The trailing CRC did not match the computed CRC (corrupt frame).
    BadCrc { expected: u8, found: u8 },
}

/// Decode exactly one MSP frame from the START of `buf` (v1 or v2, auto-detected
/// from the `$M` / `$X` preamble). Returns the frame and the number of bytes it
/// consumed, so a caller streaming from the socket can advance past it. Verifies
/// the CRC and fails closed on a mismatch — a corrupt FC reply must never be
/// handed on as if it were valid (Rule 44: a status surface reports verified
/// data).
pub fn decode_frame(buf: &[u8]) -> Result<(MspFrame, usize), MspDecodeError> {
    if buf.len() < 3 {
        return Err(MspDecodeError::TooShort);
    }
    if buf[0] != b'$' {
        return Err(MspDecodeError::BadPreamble);
    }
    match buf[1] {
        b'M' => decode_v1(buf),
        b'X' => decode_v2(buf),
        _ => Err(MspDecodeError::BadPreamble),
    }
}

fn decode_v1(buf: &[u8]) -> Result<(MspFrame, usize), MspDecodeError> {
    // '$' 'M' dir size cmd payload... crc  -> min 6 bytes (empty payload)
    if buf.len() < 6 {
        return Err(MspDecodeError::TooShort);
    }
    let dir = buf[2];
    let size = buf[3] as usize;
    let cmd = buf[4] as u16;
    let total = 6 + size;
    if buf.len() < total {
        return Err(MspDecodeError::TooShort);
    }
    let payload = &buf[5..5 + size];
    let mut crc = buf[3] ^ buf[4];
    for &b in payload {
        crc ^= b;
    }
    let found = buf[total - 1];
    if crc != found {
        return Err(MspDecodeError::BadCrc {
            expected: crc,
            found,
        });
    }
    Ok((
        MspFrame {
            cmd,
            dir,
            payload: payload.to_vec(),
            version: 1,
        },
        total,
    ))
}

fn decode_v2(buf: &[u8]) -> Result<(MspFrame, usize), MspDecodeError> {
    // '$' 'X' dir flag cmd_lo cmd_hi size_lo size_hi payload... crc -> min 9
    if buf.len() < 9 {
        return Err(MspDecodeError::TooShort);
    }
    let dir = buf[2];
    let flag = buf[3];
    let cmd = u16::from_le_bytes([buf[4], buf[5]]);
    let size = u16::from_le_bytes([buf[6], buf[7]]) as usize;
    let total = 9 + size;
    if buf.len() < total {
        return Err(MspDecodeError::TooShort);
    }
    let payload = &buf[8..8 + size];
    let mut crc = crc8_dvb_s2_buf(&buf[3..8]); // flag + cmd + size
    let _ = flag; // documented; covered by the header slice above
    for &b in payload {
        crc = crc8_dvb_s2(crc, b);
    }
    let found = buf[total - 1];
    if crc != found {
        return Err(MspDecodeError::BadCrc {
            expected: crc,
            found,
        });
    }
    Ok((
        MspFrame {
            cmd,
            dir,
            payload: payload.to_vec(),
            version: 2,
        },
        total,
    ))
}

/// Map a bipolar stick axis (roll / pitch / yaw), `-1.0..=1.0`, to a PWM value
/// `1000..=2000` with centre `1500`. Out-of-range inputs clamp. This matches the
/// GCS `bipolarToPwm` mapping so a companion and the GCS scale sticks identically.
pub fn bipolar_to_pwm(v: f32) -> u16 {
    let clamped = v.clamp(-1.0, 1.0);
    (1500.0 + clamped * 500.0).round() as u16
}

/// Map a throttle input, `0.0..=1.0`, to a PWM value `1000..=2000` with idle at
/// `1000` (NOT centre — a throttle at rest is idle, not mid-stick). Matches the
/// GCS `throttleToPwm` mapping.
pub fn throttle_to_pwm(t: f32) -> u16 {
    let clamped = t.clamp(0.0, 1.0);
    (1000.0 + clamped * 1000.0).round() as u16
}

/// Encode an `MSP_SET_RAW_RC` command carrying `channels` PWM microsecond values
/// as little-endian `u16`s. `v2` selects the framing (v2 is the safe default for
/// a modern FC; v1 is offered for a legacy target). A v1 encode of more than 127
/// channels overflows the byte size field and returns `None`.
pub fn set_raw_rc(channels: &[u16], v2: bool) -> Option<Vec<u8>> {
    let mut payload = Vec::with_capacity(channels.len() * 2);
    for &ch in channels {
        payload.extend_from_slice(&ch.to_le_bytes());
    }
    if v2 {
        Some(encode_v2(MSP_SET_RAW_RC, &payload))
    } else {
        encode_v1(MSP_SET_RAW_RC, &payload)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    // ---- Golden byte vectors: the codec is the interop boundary, so pin the
    // exact bytes rather than only round-tripping (a symmetric bug in encode +
    // decode would pass a round-trip and still be wrong on the wire).

    #[test]
    fn crc8_dvb_s2_matches_known_values() {
        // DVB-S2 CRC8 of a single 0x00 is 0x00; of {0x00,0x00} is 0x00.
        assert_eq!(crc8_dvb_s2_buf(&[0x00]), 0x00);
        // A single 0x01 through the poly 0xD5.
        assert_eq!(crc8_dvb_s2_buf(&[0x01]), 0xD5);
    }

    #[test]
    fn v1_msp_status_request_golden() {
        // MSP_STATUS (101) with no payload: $ M < 00 65 crc, crc = 0x00 ^ 0x65.
        let frame = encode_v1(101, &[]).unwrap();
        assert_eq!(frame, vec![b'$', b'M', b'<', 0x00, 0x65, 0x65]);
    }

    #[test]
    fn v1_with_payload_golden() {
        // cmd 200, payload [0xE8,0x03] (=1000 LE). size=2.
        // crc = 2 ^ 200 ^ 0xE8 ^ 0x03.
        let frame = encode_v1(200, &[0xE8, 0x03]).unwrap();
        let crc = 2u8 ^ 200 ^ 0xE8 ^ 0x03;
        assert_eq!(frame, vec![b'$', b'M', b'<', 0x02, 0xC8, 0xE8, 0x03, crc]);
    }

    #[test]
    fn v1_rejects_oversize() {
        assert!(encode_v1(256, &[]).is_none()); // cmd > 255
        assert!(encode_v1(1, &vec![0u8; 256]).is_none()); // payload > 255
    }

    #[test]
    fn v2_msp_status_request_golden() {
        // $ X < flag=00 cmd=65 00 size=00 00 crc. The CRC 0xCA is from an
        // INDEPENDENT CRC8/DVB-S2 oracle (verified against the canonical
        // "123456789" -> 0xBC check value), not this module's own function, so a
        // symmetric encode/crc bug cannot pass.
        let frame = encode_v2(101, &[]);
        assert_eq!(
            frame,
            vec![b'$', b'X', b'<', 0x00, 0x65, 0x00, 0x00, 0x00, 0xCA]
        );
    }

    #[test]
    fn set_raw_rc_v2_golden() {
        // Two channels at 1500 (0xDC 0x05) each. cmd=200 (0xC8 0x00), size=4.
        // Full frame incl. the CRC 0xB4 from the independent oracle above.
        let bytes = set_raw_rc(&[1500, 1500], true).unwrap();
        assert_eq!(
            bytes,
            vec![b'$', b'X', b'<', 0x00, 0xC8, 0x00, 0x04, 0x00, 0xDC, 0x05, 0xDC, 0x05, 0xB4]
        );
    }

    #[test]
    fn decode_roundtrips_v1_and_v2() {
        for &v2 in &[false, true] {
            let bytes = set_raw_rc(&[1000, 1500, 2000, 1500], v2).unwrap();
            let (frame, consumed) = decode_frame(&bytes).unwrap();
            assert_eq!(consumed, bytes.len());
            assert_eq!(frame.cmd, MSP_SET_RAW_RC);
            assert_eq!(frame.dir, DIR_TO_FC);
            assert_eq!(frame.version, if v2 { 2 } else { 1 });
            // Payload decodes back to the channel values.
            let chans: Vec<u16> = frame
                .payload
                .chunks_exact(2)
                .map(|c| u16::from_le_bytes([c[0], c[1]]))
                .collect();
            assert_eq!(chans, vec![1000, 1500, 2000, 1500]);
        }
    }

    #[test]
    fn decode_rejects_a_corrupt_crc() {
        let mut bytes = set_raw_rc(&[1500], true).unwrap();
        let last = bytes.len() - 1;
        bytes[last] ^= 0xFF; // corrupt the CRC
        assert!(matches!(
            decode_frame(&bytes),
            Err(MspDecodeError::BadCrc { .. })
        ));
        // v1 too.
        let mut v1 = set_raw_rc(&[1500], false).unwrap();
        let l = v1.len() - 1;
        v1[l] ^= 0xFF;
        assert!(matches!(
            decode_frame(&v1),
            Err(MspDecodeError::BadCrc { .. })
        ));
    }

    #[test]
    fn decode_reports_short_and_bad_preamble() {
        assert_eq!(decode_frame(b"$"), Err(MspDecodeError::TooShort));
        assert_eq!(
            decode_frame(&[b'$', b'Z', b'<', 0, 0, 0]),
            Err(MspDecodeError::BadPreamble)
        );
        // A v2 frame missing its payload+crc is short, not corrupt.
        assert_eq!(
            decode_frame(&[b'$', b'X', b'<', 0, 0xC8, 0, 4, 0]),
            Err(MspDecodeError::TooShort)
        );
    }

    #[test]
    fn stick_scaling_matches_the_gcs() {
        // Bipolar: centre, extremes, clamp.
        assert_eq!(bipolar_to_pwm(0.0), 1500);
        assert_eq!(bipolar_to_pwm(1.0), 2000);
        assert_eq!(bipolar_to_pwm(-1.0), 1000);
        assert_eq!(bipolar_to_pwm(2.0), 2000); // clamp high
        assert_eq!(bipolar_to_pwm(-2.0), 1000); // clamp low
        assert_eq!(bipolar_to_pwm(0.5), 1750);
        // Throttle: idle at 0, NOT centre. A throttle scaled to mid-stick would
        // arm a motor at half power the instant the channel opened.
        assert_eq!(throttle_to_pwm(0.0), 1000);
        assert_eq!(throttle_to_pwm(1.0), 2000);
        assert_eq!(throttle_to_pwm(0.5), 1500);
        assert_eq!(throttle_to_pwm(-0.1), 1000); // clamp
        assert_eq!(throttle_to_pwm(1.1), 2000); // clamp
    }
}
